File: server/services/aiService.js
/**
 * AI Service
 * Provides question rewriting for improved RAG retrieval
 */

import { hfClient } from "../config/ai.js";

/**
 * Rewrites follow-up questions into standalone agricultural queries
 * Uses LLM to extract crop and topic context from conversation history
 * @param {string} question - User's follow-up question
 * @param {string} history - Recent conversation history for context
 * @returns {Promise<string>} Rewritten standalone question
 */
export async function rewriteQuestion(question, history) {
    const prompt = `
You rewrite follow-up questions into standalone agriculture questions.

Conversation:
${history}

Follow-up question:
${question}

Rewrite it as a complete standalone question.
- Mention crop explicitly
- Mention topic (fertilizer, pest, irrigation, etc)
- Do NOT answer
`;

    const res = await hfClient.chatCompletion({
        model: "meta-llama/Meta-Llama-3.1-8B-Instruct",
        messages: [{ role: "user", content: prompt }],
        temperature: 0,
        max_tokens: 80,
    });

    return res.choices[0].message.content.trim();
}

File: backend/services/ai_service.py
from core.ai import LLMProvider, get_llm_provider
from core.config import get_settings


async def rewrite_question(
    question: str,
    *,
    session_context: str,
    history: str,
) -> str:
    """Rewrite a follow-up into a standalone retrieval query (retrieval only)."""
    prompt = f"""You rewrite follow-up questions about a financial document into standalone search queries.

SESSION CONTEXT:
{session_context}

CONVERSATION:
{history}

FOLLOW-UP QUESTION:
{question}

Rewrite as one complete standalone question suitable for document search.
- Include company, metric, and reporting period from session context when implied
- If the question is already standalone, return it unchanged
- Do NOT answer the question
- Keep financial terminology precise
- Output only the rewritten question, nothing else"""

    settings = get_settings()
    llm: LLMProvider = get_llm_provider()
    rewritten = await llm.chat_completion(
        model=settings.REWRITE_MODEL,
        messages=[{"role": "user", "content": prompt}],
        temperature=0,
        max_tokens=settings.REWRITE_MAX_TOKENS,
    )
    return rewritten.strip().strip('"') or question
